README.md
![logo](https://raw.githubusercontent.com/megalodon-chess/megalodon/main/logo/logo_widescreen_light.png)

# Megalodon

![license](https://img.shields.io/github/license/megalodon-chess/megalodon)
![issues](https://img.shields.io/github/issues/megalodon-chess/megalodon)
![pulls](https://img.shields.io/github/issues-pr/megalodon-chess/megalodon)
![build](https://img.shields.io/github/workflow/status/megalodon-chess/megalodon/CI)
![release](https://img.shields.io/github/v/release/megalodon-chess/megalodon)
![stars](https://img.shields.io/github/stars/megalodon-chess/megalodon)
![forks](https://img.shields.io/github/forks/megalodon-chess/megalodon)
![size](https://img.shields.io/github/repo-size/megalodon-chess/megalodon)

UCI chess engine.

Play against Megalodon on [Lichess][lichess]!

[Version 1.0.0][latest] was recently released!

The source code of Megalodon is completely original. It is not a derivative or port of any other program.
We used articles from [chessprogramming.org](https://chessprogramming.org) and ideas from other engines.

## Compiling

### Compiling with automated Bash Script

Requires g++, CMake, and GNU Make.

``` bash
git clone https://github.com/megalodon-chess/megalodon.git
cd ./megalodon/
./build.sh
./build/Megalodon
```

### Compiling with CMake and GNU Make

Requires g++, CMake, and GNU Make.

``` bash
git clone https://github.com/megalodon-chess/megalodon.git
cd ./megalodon/
mkdir ./build/
cd ./build/
cmake ..
make -j
./Megalodon
```

### Compiling Manually

Requires g++.

``` bash
git clone https://github.com/megalodon-chess/megalodon.git
cd ./megalodon/src/
g++ -pthread -Ofast *.cpp -o Megalodon
./Megalodon
```

## License

Everything in the repository is licensed as GNU GPL v3 license.

If you would like to use the logo, we request that you also include a message
saying that your project isn't related to Megalodon.

## The Name

The name Megalodon comes from a species of shark that lived millions of years ago.

## Documentation

For extra documentation, please read the [docs][docs].

[lichess]: https://lichess.org/@/megalodon-chess
[docs]: https://megalodon-chess.github.io/megalodon/
[latest]: https://github.com/megalodon-chess/megalodon/releases/latest


docs/releases.md
# Releases

This page contains release info.

## [Megalodon 1.0.0][v1.0.0]

Megalodon is a UCI chess engine with features like alpha-beta search, evaluation functions, and transposition tables.
Version 1.0.0 plays at an estimated 1600 ELO on Lichess. View it's profile [here](https://lichess.org/@/megalodon-chess).

### Features
* Bitboard move generation
* Minimax search with alpha-beta pruning
* Iterative deepening
* Evaluation on material, pawns, knights, queens, and kings.
* Transposition table
* Zobrist hashing
* UCI protocol

### Speed Stats
Tested on `Intel i7-9700K 3.6GHz` at starting chess position.
* Move generation at depth 6: 119,060,324 nodes, 20.8 million NPS
* Search for 2 seconds: Depth 5, 165,804 nodes, 565,884 nps
* Search to depth 6: 2,036,286 nodes, 566,736 nps, 3,592 ms

### UCI Options
Use these options to customize Megalodon's behavior.

#### Hash
This is the size of Megalodon's internal table, in megabytes. The longer you wish to run a search, the larger this value should be. The default value of 256 can hold 89 million positions, which is good for more than one hour of analysis.

#### EvalMaterial, EvalPawnStruct, EvalSpace, EvalKnights, EvalRooks, EvalQueens, EvalKings
These are weights (percent) of each evaluation category. You can use these to change Megalodon's behavior.

### Goals

#### Goals set for v1.0.0

- [x] ELO: 1200
- [x] Nodes per second: 30000
- [x] Bitboard move generation
- [x] Breadth first search (kind of)
- [x] Minimax move picking with alpha-beta pruning
- [ ] Prune lowest N% branches (late move reductions)

#### Goals for v2.0.0

* 2000 ELO
* 1 million NPS
* Late move reductions and null move pruning
* Monte carlo tree search
* More customization options
* Multithreaded search
* Syzygy tablebase support (hopefully)
* Neural networks (hopefully)
* Variants (hopefully)

### Changes from v0.4.3
* Small optimizations
* Bug fixes
* Hash table capacity was increased by 60%

## [Megalodon 0.4.3][v0.4.3]

This release contains a transposition table. It has a default size of 256 MB, and stores about 50 million positions.
Consider increasing this size for a deep search.

It plays almost 300 ELO stronger than the previous release.
This is due to a much deeper search in most positions.

Bench: 765969

## [Megalodon 0.4.2][v0.4.2]

This release features improved evaluation functions.

New evaluation categories include queens, rooks, pawn attacks, and weights.

## [Megalodon 0.4.1][v0.4.1]

This release contains many bug fixes.
The move generation is perfect when compared to [these results](https://www.chessprogramming.org/Perft_Results).

50 move draw has been implemented.
There are many projects going on about search, multithreading, eval, and Monte Carlo Tree Search.
Those may be merged shortly.

Bench: 2196682

## [Megalodon 0.4.0][v0.4.0]

This release contains many bug fixes in move generation.
Thanks to Guenther Simon (@rwbc) for reporting bugs.

The speed of everything was increased by about 5 times!
This was achieved by adding the `-O3` flag when compiling, which allows g++ to do optimization.

The v0.4.x releases will be focused on fixing bugs and evaluation algorithms.

## [Megalodon 0.3.2][v0.3.2]

There are many bug fixes in this release.

A new queen checkmate algorithm was added, and Megalodon will now use that algorithm instead of a search in queen endgames.

## [Megalodon 0.3.1][v0.3.1]

This release has move generation improvements.
The new algorithms reach 4.5 million nps, which is mostly due to using arrays instead of vectors.

There are also small evaluation weight changes.

The search function also runs in a different thread now, so the search can be stopped at any time with `stop`.

Megalodon plays at an estimated 1000 ELO.

## [Megalodon 0.3.0][v0.3.0]

This release features very fast move generation algorithms.
The move generation searches at 3.5 million NPS, a 50% improvement on 2.4 million in v0.2.6
The accuracy was also greatly improved, and agrees with the [Chess] module in python.

There are also minor evaluation changes.

[chess]: https://pypi.org/project/chess/

## [Megalodon 0.2.6][v0.2.6]

This release features many evaluation categories, including pawn structure, knights, kings, and space.

You can tune the weights with UCI options.

We will also start hosting a Sharktest server in a few days.

This allows you to use your CPU power to tune Megalodon's evaluation.

## [Megalodon 0.2.5][v0.2.5]

**THIS RELEASE CONTAINS BUGS**

This version is more than four times faster than the previous release.

There are also minor bug fixes, such as increasing the `nodes` max size.

The big speed increase is due to directly calling `eval` instead of `search` when there is one depth left.

The time management algorithm was also improved, and Megalodon will usually not time out or think very long.

You can also set when alpha-beta values are passed, and their margin through `setoption`.

## [Megalodon 0.2.4][v0.2.4]

This release features a hash table and searches a principal variation.

The hash algorithm currently has about 1 collision every 10000 hashes, so it is extremely unreliable.
We recommend not using the hash table in this version.

To toggle hash table on and off:

```
setoption name UseHashTable value true
setoption name UseHashTable value false
```

## [Megalodon 0.2.3][v0.2.3]

This release features improved search and move generation.

The evaluation has been reduced to material only, and it will be developed in v0.3.x

The move generation reaches 900k nodes per second, a 20% improvement from the previous version.

Iterative Deepening has been implemented into the new search algorithm, but there are no
improvements over single depth alpha-beta yet.

## [Megalodon 0.2.2][v0.2.2]

This release contains evaluation improvements and legal move bug fixes.

## [Megalodon 0.2.1][v0.2.1]

This release contains many bug fixes and evaluation improvements.


src/uci.cpp
//
//  Megalodon
//  UCI chess engine
//  Copyright the Megalodon developers
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

#include <iostream>
#include <vector>
#include <string>
#include <thread>
#include "utils.hpp"
#include "bitboard.hpp"
#include "search.hpp"
#include "options.hpp"
#include "eval.hpp"
#include "perft.hpp"
#include "hash.hpp"
#include "endgame.hpp"

using std::cin;
using std::cout;
using std::endl;
using std::vector;
using std::string;
using std::to_string;

const vector<string> GREETINGS = {"Hello!", "Lets play!", "Are you ready for a game?"};
const vector<string> WINNING = {"Looks like I'm playing well!", "Wow!", "This is great!"};
const vector<string> LOSING = {"Oh no!", "I blundered.", "Nice play!", "Great job!", "*sigh*. You're good."};
const vector<string> GAME_END = {"Good game!", "I look forward to playing again.", "Want to play another one?", "Rematch?"};


Position parse_pos(const vector<string>& parts) {
    if (parts[1] == "startpos") {
        Position pos = Bitboard::startpos();
        if (parts.size() > 3 && parts[2] == "moves") {
            for (char i = 3; i < parts.size(); i++) {
                pos = Bitboard::push(pos, parts[i]);
            }
        }
        return pos;
    } else if (parts[1] == "fen") {
        string fen;
        for (char i = 2; i < 8; i++) {
            fen += parts[i];
            fen += " ";
        }
        Position pos = Bitboard::parse_fen(fen);

        if (parts.size() > 9 && parts[8] == "moves") {
            for (char i = 9; i < parts.size(); i++) {
                pos = Bitboard::push(pos, parts[i]);
            }
        }
        return pos;
    }
    return Position();
}


void print_legal_moves(const Position& pos) {
    const vector<Move> moves = Bitboard::legal_moves(pos, Bitboard::attacked(pos, !pos.turn));
    cout << moves.size() << endl;
    for (const auto& m: moves) cout << Bitboard::move_str(m) << "\n";
}

void chat(const Options& options, const bool& turn, const int& movect, const float& score, const float& prev_score) {
    return;  //! CHAT IS DISABLED
    if (movect == 0) cout << "info string " << rand_choice(GREETINGS) << endl;
    else if (turn  && (score > (prev_score+1.5))) cout << "info string " << rand_choice(WINNING) << endl;
    else if (!turn && (score < (prev_score-1.5))) cout << "info string " << rand_choice(WINNING) << endl;
    else if (turn  && (score < (prev_score-1.5))) cout << "info string " << rand_choice(LOSING)  << endl;
    else if (!turn && (score > (prev_score+1.5))) cout << "info string " << rand_choice(LOSING)  << endl;
}

float go(const Options& options, const Position& pos, const vector<string>& parts, const float& prev_eval, bool& searching) {
    int mode = 0;
    int depth = 99;
    double movetime;
    const int total = Eval::total_mat(pos);
    float wtime = 0, btime = 0, winc = 0, binc = 0;
    bool infinite = false;
    for (char i = 0; i < parts.size()-1; i++) {
        if (parts[i] == "depth") {
            mode = 1;
            depth = std::stoi(parts[i+1]);
        } else if (parts[i] == "wtime") {
            mode = 2;
            wtime = std::stof(parts[i+1]) / 1000;
        } else if (parts[i] == "btime") {
            mode = 2;
            btime = std::stof(parts[i+1]) / 1000;
        } else if (parts[i] == "winc") {
            mode = 2;
            winc = std::stof(parts[i+1]) / 1000;
        } else if (parts[i] == "binc") {
            mode = 2;
            binc = std::stof(parts[i+1]) / 1000;
        } else if (parts[i] == "movetime") {
            mode = 3;
            movetime = std::stof(parts[i+1]) / 1000;
        } else if (parts[i] == "infinite") {
            infinite = true;
        }
    }

    if (mode == 0) {
        depth = 99;
        movetime = 10000000;  // About 100 days
    } else if (mode == 1) {
        movetime = 10000000;
    } else if (mode == 2) {
        depth = 99;
        if (pos.turn) movetime = Search::move_time(options, pos, wtime, winc);
        else          movetime = Search::move_time(options, pos, btime, binc);
    }
    if (mode == 2) movetime /= 1.5;

    searching = true;
    const SearchInfo result = Search::search(options, pos, depth, movetime, infinite, searching, (mode==2));
    cout << "bestmove " << Bitboard::move_str(result.pv[0]) << endl;

    chat(options, pos.turn, pos.move_cnt, result.score, prev_eval);
    return result.score;
}

void perft(const Options& options, const Position& pos, const int& depth) {
    const vector<Move> moves = Bitboard::legal_moves(pos, Bitboard::attacked(pos, !pos.turn));
    const double start = get_time();
    long long nodes = 0;

    if (moves.size() > 0) {
        int move_num = 1;
        for (const auto& move: moves) {
            const Position new_pos = Bitboard::push(pos, move);
            const long long curr_nodes = Perft::movegen(new_pos, depth-1);
            nodes += curr_nodes;
            cout << "info currmove " << Bitboard::move_str(move) << " currmovenumber " << move_num << " nodes " << curr_nodes << endl;
            move_num++;
        }
    }

    const double elapse = get_time() - start + 0.001;  // Add 1 ms to prevent divide by 0
    cout << "info depth " << depth << " nodes " << nodes << " nps " << (long long)(nodes/elapse) << " time " << (long long)(elapse*1000) << endl;
}

void perft_hash(const Options& options, const Position& pos, const int& knodes) {
    const double time = Perft::hash_perft(pos, knodes);
    cout << "info nodes " << 1000*knodes << " nps " << (int)(knodes*1000/time) << " time " << (int)(time*1000) << endl;
}

void perft_eval(const Options& options, const Position& pos, const int& knodes) {
    const double time = Perft::eval_perft(options, pos, knodes);
    cout << "info nodes " << 1000*knodes << " nps " << (int)(knodes*1000/time) << " time " << (int)(time*1000) << endl;
}

void perft_push(const Position& pos, const int& knodes) {
    const double time = Perft::push_perft(pos, knodes);
    cout << "info nodes " << 1000*knodes << " nps " << (int)(knodes*1000/time) << " time " << (int)(time*1000) << endl;
}

void bench() {
    const Options options;
    constexpr char num_pos = 20;
    constexpr char depth = 4;
    const string fens[num_pos] = {
        "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
        "r2q1rk1/ppp2pp1/2np1n1p/2b1p3/2B1P1b1/2NPBN2/PPPQ1PPP/R3R1K1 b Qq - 0 1",
        "6k1/5ppp/8/8/8/8/5PPP/6K1 w - - 0 1",
        "8/8/NnNnNnNn/8/8/8/8/2k2K2 b - - 0 1",
        "7k/p6P/p6P/p6P/p6P/p6P/p6P/K7 w - - 0 1",
        "8/8/8/8/8/2k5/8/1QK5 w - - 0 1",
        "4k3/p7/Pp6/1Pp1p3/2PpPp1p/3P1PpP/6P1/3K4 w - - 0 1",
        "8/1k6/3K4/2P5/8/8/8/8 w - - 0 1",
        "r2q1rk1/pbp1bpp1/1pn1pn1p/3p4/2PP3B/1PN1PN2/P3BPPP/R2Q1RK1 b Qq - 0 1",
        "2k4r/1pp5/p7/2p1N3/3n1P2/7P/6P1/R5K1 b - - 0 1",
        "7k/p7/8/8/8/8/P7/7K b - - 0 1",
        "8/8/8/3b4/8/3Bk3/1BK2b2/8 b - - 0 1",
        "8/3qPK2/8/8/8/4k3/8/8 w - - 0 1",
        "8/7P/5K2/8/8/4k3/p7/8 w - - 0 1",
        "r5k1/2p2ppp/pp2n3/8/P7/1B1P3P/1PP2PP1/3R1RK1 b - - 0 1",
        "4k3/pppppppp/8/8/8/8/PPPPPPPP/4K3 b - - 0 1",
        "r1r1r1r1/k7/8/8/8/8/1K6/1R1R1R1R w - - 0 1",
        "K7/8/8/3P4/4p3/8/8/7k b - - 0 1",
        "8/kpP3p1/p2n2P1/3Q3P/5q2/P7/KP6/8 w - - 0 1",
        "8/8/8/8/8/k7/8/K7 w - - 0 1",
    };

    U64 nodes = 0;
    const double start = get_time();
    bool searching = true;

    for (char i = 0; i < num_pos; i++) {
        cout << "Position " << i+1 << " of " << +num_pos << endl;
        cout << "Fen: " << fens[i] << endl;

        const Position pos = Bitboard::parse_fen(fens[i]);
        const SearchInfo result = Search::search(options, pos, depth, 10000, false, searching, false);
        nodes += result.nodes;
    }

    const double elapse = get_time() - start;
    const int nps = nodes / elapse;

    cout << "\nBenchmark results:" << endl;
    cout << "Nodes: " << nodes << endl;
    cout << "NPS: " << nps << endl;
    cout << "Time: " << elapse << " seconds" << endl;
}

void parse_command(Options& options, Position& pos, float& prev_eval, bool& searching, const vector<string>& parts, const bool& arg=false) {
    if (parts[0] == "quit") return;
    else if (parts[0] == "clear") {
        string str;
        for (const char& i: {27, 91, 51, 74, 27, 91, 72, 27, 91, 50, 74}) str += string(1, i);
        cout << str << std::flush;
    }
    else if (parts[0] == "isready") cout << "readyok" << endl;
    else if (parts[0] == "uci") {
        cout << "id name Megalodon" << "\n";
        cout << "id author Megalodon Developers" << "\n";

        cout << "option name Hash type spin default 256 min 1 max 65536" << "\n";

        cout << "option name EvalMaterial type spin default 100 min 0 max 1000" << "\n";
        cout << "option name EvalPawnStruct type spin default 100 min 0 max 1000" << "\n";
        cout << "option name EvalSpace type spin default 100 min 0 max 1000" << "\n";
        cout << "option name EvalKnights type spin default 100 min 0 max 1000" << "\n";
        cout << "option name EvalRooks type spin default 100 min 0 max 1000" << "\n";
        cout << "option name EvalQueens type spin default 100 min 0 max 1000" << "\n";
        cout << "option name EvalKings type spin default 100 min 0 max 1000" << "\n";

        cout << "uciok" << endl;
    }
    else if (parts[0] == "setoption") {
        const string name = parts[2];
        const string value = parts[4];

        if (name == "Hash") {
            options.Hash = std::stoi(value);
            options.set_hash();
        }

        else if (name == "EvalMaterial")   options.EvalMaterial   = std::stof(value)/100;
        else if (name == "EvalPawnStruct") options.EvalPawnStruct = std::stof(value)/100;
        else if (name == "EvalSpace")      options.EvalSpace      = std::stof(value)/100;
        else if (name == "EvalKnights")    options.EvalKnights    = std::stof(value)/100;
        else if (name == "EvalRooks")      options.EvalRooks      = std::stof(value)/100;
        else if (name == "EvalQueens")     options.EvalQueens     = std::stof(value)/100;
        else if (name == "EvalKings")      options.EvalKings      = std::stof(value)/100;

        else std::cerr << "Unknown option: " << name << endl;
    }

    else if (parts[0] == "d") cout << Bitboard::board_str(pos) << endl;
    else if (parts[0] == "hash") {
        if (parts.size() == 1) {
            cout << Hash::hash(pos) << endl;
        } else if (parts[1] == "perft" && parts.size() >= 2) {
            perft_hash(options, pos, std::stoi(parts[2]));
        }
    }
    else if (parts[0] == "eval") {
        if (parts.size() == 1) {
            const U64 attacked = Bitboard::attacked(pos, !pos.turn);
            Eval::eval(options, pos, Bitboard::legal_moves(pos, attacked), 0, attacked, true);
        } else if (parts[1] == "perft" && parts.size() >= 2) {
            perft_eval(options, pos, std::stoi(parts[2]));
        }
    }
    else if (parts[0] == "bench") bench();
    else if (parts[0] == "legalmoves") print_legal_moves(pos);
    else if (parts[0] == "pushperft") perft_push(pos, std::stoi(parts[1]));
    else if (parts[0] == "eg") cout << Endgame::eg_type(pos) << endl;

    else if (parts[0] == "ucinewgame") {
        pos = parse_pos({"position", "startpos"});
        prev_eval = 0;
    }
    else if (parts[0] == "position") pos = parse_pos(parts);
    else if (parts[0] == "go") {
        if (parts.size() > 1 && parts[1] == "perft") perft(options, pos, std::stoi(parts[2]));
        else {
            options.clear_hash();
            searching = true;
            if (arg) go(options, pos, parts, prev_eval, searching);
            else std::thread(go, options, pos, parts, prev_eval, std::ref(searching)).detach();
        }
    }
    else if (parts[0] == "stop") searching = false;
    else if (parts.size() > 0) std::cerr << "Unknown command: " << join(" ", parts) << endl;
}


int loop() {
    string cmd;
    Options options;
    Position pos = parse_pos({"position", "startpos"});
    float prev_eval = 0;
    bool searching = false;

    while (getline(cin, cmd)) {
        cmd = strip(cmd);
        const vector<string> parts = split(cmd, " ");
        parse_command(options, pos, prev_eval, searching, parts);
    }

    searching = false;
    delete[] options.hash_table;
    return 0;
}
